Fix thread safety issues in SwitchProducer ProductResolvers

makortel opened the pull request:
#### PR description:

This PR fixes three thread safety issues in SwitchProducer ProductResolvers that I realized while addressing #28680
- `SwitchBaseProductResolver::prefetchRequested_` is now re-set to `false` in `resetProductData_()`
   * Current behavior (without re-setting) could lead to starvation if there are many consumers (in Tasks) for the SwitchProducer products. This indeed happens with enhanced tests (multiple consumers, multiple threads)
- Call `ProductData::unsafe_setWrapper()` in the prefetching phase as suggested by @Dr15Jones in #28680 
   * The ordering of the tasks ensures that exactly one task is accessing the `ProductData` at the time of the call, after which the tasks for the product consumers are spawned.
- `SwitchBaseProductResolver::waitingTasks_` is reset in `resetProductData_()`
  * Current behavior (without reset) leads to product consumer tasks being spawned too early from 2nd event onwards

In addition this PR
- Moves `putProduct_()` from `SwitchBaseProductResolver` to `SwitchProducerProductResolver`
   * The `putProduct_()` should never be called from `SwitchAliasProductResolver` (and now throws an exception)
- Moves the call to `updateProvenance()` to a proper place in `SwitchProducerProductResolver`
   * The earlier placement looked suspicious (why update if got an exception), and the problem in the provenance tracking was confirmed by enhancing the tests to check the provenance information also for the test configuration where the SwitchProcucer is in a Path

Fixes #28680.

#### PR validation:

Code compiles, unit tests run.

makortel committed:
Reset SwitchBaseProductResolver::prefetchRequested_

makortel committed:
Move putProduct_() from SwitchBaseProductResolver to SwitchProducerProductResolver

putProduct_() should never be called fro SwitchAliasProductResolver,
which means that the infrastructure to support putProduct_() is not
common for the two. Moving the function to
SwitchProducerProductResolver makes the intentions more clear.

Also remove a useless return from updateProvenance().

makortel committed:
Call unsafe_setWrapper() in prefetching phase

makortel committed:
Reset WaitingTaskList

makortel committed:
Run SwitchProducer tests in multiple threads, add multiple consumers

makortel committed:
Call updateProvenance() properly in SwitchProducerProductResolver

Found by checking provenance from SwitchProducer-in-Path as well

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13257

- This PR adds an extra 40KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13257/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13257/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13258

- This PR adds an extra 40KB to repository

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

FWCore/Framework
FWCore/Integration


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos, @silviodonato you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

FWCore/Framework
FWCore/Integration


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@wddgit this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos, @silviodonato you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on FWCore/Framework/src/ProductResolvers.h:
```diff
@@ -40,6 +40,7 @@ namespace edm {
 
     // Give AliasProductResolver and SwitchBaseProductResolver access by moving this method to public
     void resetProductData_(bool deleteEarly) override = 0;
+    virtual ProductData const& getProductData() const = 0;
```
I'm not very happy about exposing the `ProductData` this way. It is used here
https://github.com/cms-sw/cmssw/blob/9890e15de9fa531e391ffe4d0f5e6a3cfc6669c8/FWCore/Framework/src/ProductResolvers.cc#L659-L662
to set the Wrapper in the prefetching phase.

makortel commented on FWCore/Framework/src/ProductResolvers.h:
```diff
@@ -336,8 +331,17 @@ namespace edm {
+    // for filter in a Path
+    // The variable is only modified or read at times where the
+    //  framework has guaranteed synchronization between write and read
+    CMS_THREAD_SAFE mutable ProductStatus status_;
```
Contrary to my earlier thought (https://github.com/cms-sw/cmssw/pull/28667#discussion_r360576272), I believe now that this variable could stay non-atomic. The `status_` is read in `resolveProduct_()` and `putProduct_()`, and written in `putProduct_()` and `resetProductData_()`. It is not used to synchronize anything else (via memory fence). If we indeed guarantee that `resolveProduct_()` and `putProduct_()` may not be called concurrently, it should be fine as it is.

cmsbuild commented:
-1

Tested at: 9890e15de9fa531e391ffe4d0f5e6a3cfc6669c8

CMSSW: CMSSW_11_1_X_2019-12-27-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4107/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestIntegrationSwitchProducer had ERRORS

cmsbuild commented:
Comparison job queued.

makortel commented:
The failure is
```
Test provenance of merged output (Path)
...
Begin processing the 1st record. Run 1, Event 1, LumiSection 2 on stream 0 at 27-Dec-2019 21:35:13.975 CET
Begin processing the 2nd record. Run 1, Event 2, LumiSection 2 on stream 0 at 27-Dec-2019 21:35:13.977 CET
FWCore/Integration/test/SwitchProducerProvenanceAnalyzer.cc:57: void edmtest::SwitchProducerProvenanceAnalyzer::testProduct(const edm::Handle<edmtest::IntProduct>&, int, const edm::Event&) const: Assertion `productProvenance != nullptr' failed.
```
like the provenance would still not be set correctly. Somehow the failure is also in the second event (which could hint towards some other synchronization issue).

The test worked on my local tests, but after using even more threads I managed to reproduce the failure. Investigating...

makortel committed:
Add missing call to update provenance

cmsbuild commented:
The code-checks are being triggered in jenkins.

makortel commented:
Indeed one call to `updateProcenance()` was missing, reorganized the code such that the provenance is always updated together with setting the Wrapper.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13259

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #28686 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4107/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2815765
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2815418
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

cmsbuild commented:
+1
Tested at: 9484ce3fdb4491cec2f833b2ddcc32669663840e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4108/summary.html
CMSSW: CMSSW_11_1_X_2019-12-27-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4108/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2815765
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2815418
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Dr15Jones commented on FWCore/Framework/src/ProductResolvers.cc:
```diff
@@ -665,21 +648,26 @@ namespace edm {
-      status_ = defaultStatus_;
-    }
+    prefetchRequested_ = false;
+    waitingTasks_.reset();
```
If `deleteEarly == true` then neither of these items should be reset.

makortel replied:
Other ProductResolvers seem to reset these variables unconditionally
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L295-L297
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L359-L362
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L450-L452
https://github.com/cms-sw/cmssw/blob/4010528c5c1660debe651b9f69d474a7f8b5004b/FWCore/Framework/src/ProductResolvers.cc#L1026-L1032

Also, shouldn't the early deletion be done only after all consumers of the product have been run (i.e. no new prefetch requests are coming)? Ok, the current configuration-driven mechanism doesn't guarantee that, but at least in principle.

makortel replied:
The last commit makes the resets only if `deleteEarly == false`, and also adds a unit test that fails if they would not be reset.

makortel committed:
Reset prefetchRequested and WaitingTaskList only if reset does not originate from early delete

Add tests for consuming (but not reading) a product after early
deletion. The tests show that resetting these members also for
early-delete-resets leads to another round of prefetching. In case the
producer is in a Task (i.e. UnscheduledProductResolver is used for the
product), an exception gets thrown even if the consumer does not read
the product.

Also update class name name in the early deletion exception message.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28686/13263

- This PR adds an extra 52KB to repository

cmsbuild commented:
Pull request #28686 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d521f7f7a2354a38b466cb6d1f16c981a2cbecf1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4110/summary.html
CMSSW: CMSSW_11_1_X_2019-12-30-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16228b/4110/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2815513
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2815166
 - DQMHistoTests: Total skipped: 346
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos, @silviodonato (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
